modules/dists/StencilDist.chpl

```diff
@@ -1436,25 +1506,144 @@ proc _array.updateFluff() {
   _value.dsiUpdateFluff();
 }
 
-// copy over into cache
-proc StencilArr.dsiUpdateFluff() {
-  if zeroTuple(dom.fluff) then return;
+//
+// On each locale, update the local cache by fetching data from neighbors.
+//
+proc StencilArr.naiveUpdateFluff() {
   coforall i in dom.dist.targetLocDom {
     on dom.dist.targetLocales(i) {
       ref myLocDom = locArr[i].locDom;
-      forall (S, D, N, L) in zip(myLocDom.Src, myLocDom.Dest,
+      forall (S, D, N, L) in zip(myLocDom.recvSrc, myLocDom.recvDest,
           myLocDom.Neighs, myLocDom.NeighDom) {
-        if !zeroTuple(L) {
-          if !dom.dist.targetLocDom.member(i+L) && dom.periodic then
-            locArr[i].myElems[D] = locArr[N].myElems[S];
-          else if dom.dist.targetLocDom.member(N) then
-            locArr[i].myElems[D] = locArr[N].myElems[D];
+        // A source domain of size zero indicates that no communication
+        // is necessary.
+        //
+        // if "L" is zero, that indicates we are at the center of the stencil
+        // and do not need to update
+        if !isZeroTuple(L) && S.size != 0 {
+          locArr[i].myElems[D] = locArr[N].myElems[S];
         }
       }
     }
   }
 }
 
+//
+// TODO: GET or PUT the data? Does it matter for performance?
+//
+// TODO: should we avoid doing the packed transfer for dense-ish regions?
+// e.g. {1..5, 1..100} might only require 5 GETs/PUTs
+//
+// This is an optimized variant of the naiveUpdateFluff method. This variant
+// packs elements into a buffer such that we only do one GET/PUT. This is
+// most beneficial when the communicated region would require many GETs in
+// the naive approach. For example, Let's say we have a 10x10 domain with
+// a fluff of (1,1). Our wholeFluff domain would be {0..11, 0..11}. When we go
+// to communicate the region {1..10,1..1}, the naive method requires 10 GETs or
+// PUTs, whereas the packed method requires just one.
+//
+// On each locale:
+// 1) Repeat the following for each region of data that other locales will
+//    want:
+//    to fetch:
+//    a) Serialize/pack the region of data into a 1D buffer array. This is the
+//       buffer that other locales will read from.
+//    b) Write 'true' to an atomic flag on the local that intends to read from
+//       the buffer that was just populated.
+// 2) Once we have prepared our local data for reading, repeat the following
+//    for each region of data that *our* cache wants to update:
+//    a) Wait for the corresponding atomic flag to be 'true' so that we know
+//       it is safe to fetch the data from the remote locale. Reset the flag to
+//       false afterwards.
+//    b) Bulk-copy the remote buffer into a local buffer
+//    c) Copy elements from the local buffer into the cache
+//
+proc StencilArr._packedUpdate() {
```

Comment:
Consider replacing the "_" in _packedUpdate with `private` annotation, for a nicer naming convention. If so, add `private` to some other methods as well?